---
title: "Analytics 2: Introduction to the command-line for Bioinformatics"
author: "Module Coordinator Mark Dunning"
output: 
  html_document: 
    css: stylesheets/styles.css
editor_options: 
  chunk_output_type: inline
---

![](images/header-logo.png)

![](images/logo-sm.png)

## Course Content


This course offers an introduction to working with Linux. We will describe the Linux environment so that participants can start to utilize command-line tools and feel comfortable using a text-based way of interacting with a computer. We will also introduce the "Git" version control system, which is a very powerful tool to help us secure our code from data loss. Weâ€™ll cover the terminology, how you can use it for your own projects and then how you can use it to make collaboration on a project easy.

## Pre-requisites

No prior programming experience is required. Computers will be provided. You can bring your own laptop to the workshop if you wish but it will need a working wi-fi connection and you will have to let the tutor know in advance so a temporary York password can be created for you.

## Learning Outcomes

After this course you should be able to:

- Connect to a Unix / Linux system
- Navigate around a file system by issuing commands; rather than using a Desktop environment
- Move and copy files and directories within the Linux system
- Work with text files
- Run programs from the command-line
- Write short scripts to document an analysis
- Use version control (git) to track and document changes to a file

## Schedule (Provisional)

- 10:15 - 10:45 [Introduction and Course Setup](part0.nb.html)
- 10:45 - 12:00 [Introduction to the command line](https://datacarpentry.org/shell-genomics/01-introduction)
  + [Navigating Files and Directories](https://datacarpentry.org/shell-genomics/02-the-filesystem.html)
  + [Working with Files and Directories](https://datacarpentry.org/shell-genomics/03-working-with-files.html)
- 13:00 - 14:00 Introduction to command line (continued)
  + [Redirection](https://datacarpentry.org/shell-genomics/04-redirection.html)
  + [Writing Scripts and Working with Data](https://datacarpentry.org/shell-genomics/05-writing-scripts.html)
- 14:00 - 14:45 [Obtaining Sequencing data and quality assessment](part2.nb.html)
- 15:15 - end [Introduction to Version Control, and git - Presented by The Data Science Hub, University of York](https://github.com/alastair-droop/northernbug10-git-workshop)


## Optional Materials

The previous version of this workshop all had a section on using Bioinformatics workflows and pipelines; with a particular example of RNA-seq.

+ [Intro to Workflows, Pipelines and Workflow Managers](part3.nb.html)


## Unix "cheatsheet"

- [Unix cheatsheet](https://upload.wikimedia.org/wikipedia/commons/7/79/Unix_command_cheatsheet.pdf)

## Course Feedback

Please use this short form to give us feedback on the workshop. This will be used to inform the planning of future White Rose DTP training

- [Feedback form]()
